### DISCUSSION TITLE
Finding 2D CRS from 3D to consistently build a compound horizontal + vertical CRS

### DISCUSSION THREAD

**Discussion by rhugonnet** [pyproj4/pyproj#1265] [Q&A] [answered]
Hi @snowman2 and community!

I'm having issues consistently building a compound horizontal + vertical CRS in an automated manner for any DEM, which involves converting a 3D CRS to 2D. The horizontal CRS is known from the metadata, while the vertical CRS can sometimes be user-input based.

**Why the issue arose**
3D CRS seem to now be required to transform the `z` coordinate (since pyproj 3.3/PROJ 8? I didn't run into issues before and couldn't find detailed notes on that). 
For instance, converting `z` from `CRS("EPSG:4326")` does not work anymore, but it does from `CRS("EPSG:4979")` which is also `CRS("EPSG:4326").to_3d()`.

If I understood the several linked issues that I read (e.g., https://github.com/pyproj4/pyproj/issues/630, https://gis.stackexchange.com/questions/436401/transforming-epsg4979-to-given-wkt-epsg6654), this is now the expected behaviour?

**Issue**
Having 3D required for `z` transformation is fine by me. However, when trying to build consistently a `CompoundCRS` from a `VerticalCRS` or a `BoundCRS`, 3D CRS are not accepted as part of the compounds:
```python
CRS("EPSG:4979+5773")
```
```
pyproj.exceptions.CRSError: Invalid projection: EPSG:4979+5773: (Internal Proj Error: proj_create: components of the compound CRS do not belong to one of the allowed combinations of http://docs.opengeospatial.org/as/18-005r5/18-005r5.html#34)
```
while 2D work:
```python
CRS("EPSG:4326+5773")
```

I need to be able to find the 2D version of a CRS (e.g., `CRS("EPSG:4326")`) from its 3D version (e.g., `CRS("EPSG:4979")`) in an automated manner, to build a `CompoundCRS` from any 2D version of a CRS and any `VerticalCRS` or `BoundCRS`.

But I could not find a `CRS.to_2d()` function or equivalent. Is there any that I didn't find? Or maybe an obvious solution that I missed? :sweat_smile:
Thanks for the help! :)

**Comment by snowman2** [comment]
> this is now the expected behaviour?

Yes, you have to explicitly specify a 2D or 3D CRS.

> But I could not find a CRS.to_2d() function or equivalent. Is there any that I didn't find? Or maybe an obvious solution that I missed?

Normally, I would suggest using `CRS.sub_crs_list`, but that doesn't exist for the 3D Geographic CRS. I don't think there s a `to_2d` function at the moment.

@rouault do you have any suggestions?

**Reply by rouault** [reply]
> I don't think there s a `to_2d` function at the moment.

proj_crs_demote_to_2D(): https://github.com/OSGeo/PROJ/blob/285b4dcebe89cac38c69f9ce3239276950802908/src/iso19111/c_api.cpp#L4274

**Reply by snowman2** [reply]
Nice! Thanks @rouault 👍

**Reply by rhugonnet** [reply]
Amazing, thanks! Do you need a hand wrapping that into `pyproj`? I could give it a try right now.

**Reply by snowman2** [reply]
That would be great!

**Reply by snowman2** [reply]
Related: https://github.com/OSGeo/PROJ/pull/3119

**Reply by rhugonnet** [reply, marked as answer]
For the record (if someone sees this discussion): we wrapped `CRS.to_2d()` in `pyproj` in #1267.
